File: packages/kubernetypes-gen/internal/kubernetypes/type.ts
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Mateusz Paduszyński
 */
import { z } from 'zod';

export type Type =
  | StringType
  | NumberType
  | BooleanType
  | StringLiteralType
  | ObjectLiteralType
  | ArrayType
  | ReferenceType
  | IntersectionType
  | UnionType
  | NonNullType
  | InterfaceType
  | FunctionType
  | UnknownType;
export const Type: z.ZodType<Type> = z.lazy(() =>
  z.discriminatedUnion('kind', [
    StringType,
    NumberType,
    BooleanType,
    StringLiteralType,
    ObjectLiteralType,
    ArrayType,
    ReferenceType,
    IntersectionType,
    UnionType,
    NonNullType,
    InterfaceType,
    FunctionType,
    UnknownType,
  ]),
);

export type StringType = z.infer<typeof StringType>;
export const StringType = z.object({
  kind: z.literal('string'),
});

export type NumberType = z.infer<typeof NumberType>;
export const NumberType = z.object({
  kind: z.literal('number'),
});

export type BooleanType = z.infer<typeof BooleanType>;
export const BooleanType = z.object({
  kind: z.literal('boolean'),
});

export type StringLiteralType = z.infer<typeof StringLiteralType>;
export const StringLiteralType = z.object({
  kind: z.literal('string-literal'),
  raw: z.string(),
});

export type ObjectLiteralType = {
  kind: 'object-literal';
  props?: ObjectProperty[];
  indexSignature?: IndexSignature;
};
export const ObjectLiteralType = z.object({
  kind: z.literal('object-literal'),
  props: z.array(z.lazy(() => ObjectProperty)).optional(),
  indexSignature: z.lazy(() => IndexSignature).optional(),
});

export type ObjectProperty = z.infer<typeof ObjectProperty>;
export const ObjectProperty = z.object({
  name: z.string(),
  description: z.string().optional(),
  optional: z.boolean(),
  type: Type,
});

export type IndexSignature = z.infer<typeof IndexSignature>;
export const IndexSignature = z.object({
  key: Type,
  val: Type,
});

export type ArrayType = {
  kind: 'array';
  element: Type;
};
export const ArrayType = z.object({
  kind: z.literal('array'),
  element: Type,
});

export type ReferenceType = {
  kind: 'reference';
  namespace?: string;
  name: string;
  params?: Type[];
};
export const ReferenceType = z.object({
  kind: z.literal('reference'),
  namespace: z.string().optional(),
  name: z.string(),
  params: z.array(Type).optional(),
});

export type IntersectionType = {
  kind: 'intersection';
  elements: Type[];
};
export const IntersectionType = z.object({
  kind: z.literal('intersection'),
  elements: z.array(Type),
});

export type UnionType = {
  kind: 'union';
  elements: Type[];
};
export const UnionType = z.object({
  kind: z.literal('union'),
  elements: z.array(Type),
});

export type NonNullType = z.infer<typeof NonNullType>;
export const NonNullType = z.object({
  kind: z.literal('non-null'),
});

export type InterfaceType = z.infer<typeof InterfaceType>;
export const InterfaceType = z.object({
  kind: z.literal('interface'),
  methods: z.array(z.lazy(() => InterfaceMethod)).optional(),
});

export const InterfaceMethod = z.object({
  name: z.string(),
  embedded: z.boolean(),
});

export type FunctionType = z.infer<typeof FunctionType>;
export const FunctionType = z.object({
  kind: z.literal('function'),
});

export type UnknownType = z.infer<typeof UnknownType>;
export const UnknownType = z.object({
  kind: z.literal('unknown'),
});

File: packages/kubernetypes-gen/internal/kubernetypes/type.go
package kubernetypes

import (
	"encoding/json"
	"strings"
)

type Type struct {
	Kind string `json:"kind,omitempty"`
	// "kind": "reference"
	Pkg       string  `json:"-"`
	Namespace string  `json:"namespace,omitempty"`
	Name      string  `json:"name,omitempty"`
	Params    []*Type `json:"params,omitempty"`
	// "kind": "object"
	Props          []*ObjectProperty `json:"props,omitempty"`
	IndexSignature *IndexSignature   `json:"indexSignature,omitempty"`
	// "kind": "array"
	Element *Type `json:"element,omitempty"`
	// "kind": "intersection" / "union"
	Elements []*Type `json:"elements,omitempty"`
	// "kind": "string-literal"
	Raw string `json:"raw,omitempty"`
	// "kind": "interface"
	Methods []*InterfaceMethod `json:"methods,omitempty"`
}

type ObjectProperty struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Optional    bool   `json:"optional"`
	Type        *Type  `json:"type"`
}

func objectPropertyCmp(a, b *ObjectProperty) int {
	return strings.Compare(a.Name, b.Name)
}

type IndexSignature struct {
	Key *Type `json:"key"`
	Val *Type `json:"val"`
}

type InterfaceMethod struct {
	Name     string `json:"name,omitempty"`
	Embedded bool   `json:"embedded"`
}

func interfaceMethodCmp(a, b *InterfaceMethod) int {
	return strings.Compare(a.Name, b.Name)
}

var (
	stringType   = &Type{Kind: "string"}
	numberType   = &Type{Kind: "number"}
	booleanType  = &Type{Kind: "boolean"}
	functionType = &Type{Kind: "function"}
	nonNullType  = &Type{Kind: "non-null"}
	unknownType  = &Type{Kind: "unknown"}
)

func referenceType(pkg string, name, namespace string) *Type {
	return &Type{Kind: "reference", Pkg: pkg, Name: name, Namespace: namespace}
}

func objectLiteralType(props []*ObjectProperty, indexSignature *IndexSignature) *Type {
	return &Type{Kind: "object-literal", Props: props, IndexSignature: indexSignature}
}

func arrayType(element *Type) *Type {
	return &Type{Kind: "array", Element: element}
}

func intersectionType(elements []*Type) *Type {
	return &Type{Kind: "intersection", Elements: elements}
}

func unionType(elements []*Type) *Type {
	return &Type{Kind: "union", Elements: elements}
}

func stringLiteralType(value string) *Type {
	raw, _ := json.Marshal(value)
	return &Type{Kind: "string-literal", Raw: string(raw)}
}

func interfaceType(methods []*InterfaceMethod) *Type {
	return &Type{Kind: "interface", Methods: methods}
}
